### PR TITLE
VACMS-4105: Render JSD Widget on Access-Denied pages

### PR DESCRIPTION
**Pull request by ndouglas**
## Description

See #4105. 

## Testing done
Clicked a couple admin pages as both a :godmode: admin and non-admin user.

Started writing tests but ran into issues with visibility, presumably because of the delay and JS involved with Behat.

## Screenshots
Normal User:
Screen Shot 2021-02-03 at 2 18 55 PM
Admin User:
Screen Shot 2021-02-03 at 2 19 17 PM

## QA steps
As a user with non-administrative privileges:
1. Visit some path to which you shouldn't have access, like `/admin/reports/status`
1. Then validate Acceptance Criteria from issue #4105:
- [ ] JSD widget is rendered on the Access Denied page in CMS

### Definition of Done

- [ ] Documentation has been updated, if applicable.
- [ ] Tests have been added if necessary.
- [ ] Automated tests have passed.
- [ ] Code Quality Tests have passed.
- [ ] Acceptance Criteria in related issue are met.
- [ ] Manual Code Review Approved.

### Select Team for PR review

- [x] `Core Application Team`
- [ ] `Product Support Team`

### Is this PR blocked by another PR?

- [ ] `DO NOT MERGE`

### Does this PR need review from a Product Owner

- [ ] `Needs PO review`

### CMS user-facing annoucement

Is an announcement needed to let editors know of this change? 
- [ ] Yes, and it's written in issue ____ and queued for publication. 
  - [ ] Merge and ping @ rachel-kauff so she's ready to publish after deployment
- [ ] Yes, but it hasn't yet been written 
  - [ ] Don't merge yet -- ping @ rachel-kauff to prompt her to write and queue content
- [x] No announcement is needed for this code change. 
  - [x] Merge & carry on unburdened by announcements

### CODE REVIEW

File: docroot/modules/custom/va_gov_help_center/va_gov_help_center.module [hunk 1]
```diff
@@ -40,6 +40,10 @@ function va_gov_help_center_page_attachments(array &$attachments) {
     // Attach Jira Service Desk widget library to user login route.
     $attachments['#attached']['library'][] = 'va_gov_help_center/helpdesk-widget';
   }
+  elseif ($route_name === 'system.403') {
```
Comment (oksana-c) [inline docroot/modules/custom/va_gov_help_center/va_gov_help_center.module:43]: I may be totally out of style here 💃 🕺 , but can we add 403 route as OR condition in the if statement above? DRY(er)

**Review by oksana-c — COMMENTED** [review]
not a requirement (I also forgot to do it myself when created `va_gov_help_center` module), but can we move `docroot/modules/custom/va_gov_backend/src/Plugin/Block/WhatsNew.php` to `va_gov_help_center` so as to keep all help center related functionality in place?

File: docroot/modules/custom/va_gov_help_center/va_gov_help_center.module (hunk 1 above)
Comment (ndouglas) [reply docroot/modules/custom/va_gov_help_center/va_gov_help_center.module:43]: Sure.  I was thinking about writing some tests in Cypress and then refactoring to see if I could get that attachment in only one place 😄

Comment (ndouglas) [reply docroot/modules/custom/va_gov_help_center/va_gov_help_center.module:43]: I suspect we're going to add more cases as time goes by.

Comment (oksana-c) [reply docroot/modules/custom/va_gov_help_center/va_gov_help_center.module:43]: always possible that we may add more 👍

**Comment by acrollet** [comment]
🌲 🌲 🌲

**Comment by ndouglas** [comment]
(Don't worry, @acrollet, I know I'll probably be rewriting/refactoring/rm -rf'ing this after we have the spike 😁 )

**Comment by acrollet** [comment]
> (Don't worry, @acrollet, I know I'll probably be rewriting/refactoring/rm -rf'ing this after we have the spike 😁 )

nah this just means you're in charge of the spike 🧐

**Comment by ndouglas** [comment]
I still have a couple things to do on this:

- [ ] I'd like to add a couple more tests since I figured out the iframe was the issue
- [ ] move the thing to the module like Oksana said

**Comment by ndouglas** [comment]
@acrollet I'd also dig your thoughts on this:

`I am at "<page>"` is not fault-tolerant.  So if you visit `/some-page-that-doesn't-exist`, you get a 404 error and Cypress kicks up an error.  Good.

The problem is that sometimes we want to visit a page and get an error.  For instance, a 403 error, so that we can see if the JSD widget pops up.

I used `I attempt to visit "<page>"` as a fault-tolerant version of that, but I feel like there should be a better phrasing, and I'd like to get that better phrasing in place before we write a bunch of tests :D  

Thoughts?

**Comment by acrollet** [comment]
> Yeah, agreed, I'll toss that in.

**Comment by acrollet** [comment]
> I used `I attempt to visit "<page>"` as a fault-tolerant version of that, but I feel like there should be a better phrasing, and I'd like to get that better phrasing in place before we write a bunch of tests :D

`And I navigate to "<page>"`???

File: tests/behavioral/cypress/integration/common/i_should_see_element.js [hunk 2]
```diff
@@ -0,0 +1,5 @@
+import { Then } from "cypress-cucumber-preprocessor/steps";
+
+Then(`I should see element {string}`, (text) => {
```
Comment (acrollet) [inline tests/behavioral/cypress/integration/common/i_should_see_element.js]: Maybe this could be something like "I should see the element with selector {string}" to make it more self-documenting?

Comment (ndouglas) [reply tests/behavioral/cypress/integration/common/i_should_see_element.js]: Sure, sounds good to me.

File: docroot/modules/custom/va_gov_help_center/va_gov_help_center.module [hunk 3]
```diff
@@ -35,26 +34,40 @@ function va_gov_help_center_toolbar() {
...
+    'user.login',
+  ];
+  if (in_array($route_name, $unconditional_routes)) {
+    return TRUE;
+  }
+  if ($route_name === 'entity.node.canonical' && !empty($route_parameters['node']) && $node = $route_parameters['node']) {
+    // Attach Jira Service Desk widget library to all nodes of type
+    // `documentation_page` with an exception of CMS Help Center homepage -
+    // NID 6454.
+    if ($node->getType() === 'documentation_page' && $node->id() != 6454) {
```
Comment (acrollet) [inline docroot/modules/custom/va_gov_help_center/va_gov_help_center.module:68]: Since we've already got a couple of mentions of 6454 in the module should we consider setting up a CMS_HELP_CENTER_HOMEPAGE_NID constant or some such?

Comment (ndouglas) [reply docroot/modules/custom/va_gov_help_center/va_gov_help_center.module:68]: Yeah, magic numbers suck.

File: tests/behavioral/cypress/integration/common/i_am_logged_in_as_a_user_with_the_role.js [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import { Then } from "cypress-cucumber-preprocessor/steps";
+import faker from "faker";
+
+Then(`I am logged in as a user with the {string} role`, (text) => {
+  let username = faker.internet.userName();
+  let password = faker.internet.password();
+  cy.drupalAddUserWithRole(text, username, password);
```
Comment (acrollet) [inline tests/behavioral/cypress/integration/common/i_am_logged_in_as_a_user_with_the_role.js:7]: I like this, is there anything happening to remove the created user after the test is complete?

File: tests/behavioral/cypress/integration/common/i_create_a_node.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    cy.scrollTo('top');
+    cy.findAllByLabelText('Page title').type(faker.lorem.sentence(), { force: true });
+    cy.findAllByLabelText('Page introduction').type(faker.lorem.sentence(), { force: true });
+    cy.get('form.node-form').find('input#edit-submit').click();
+  }
+};
+
+Given(`I create a {string} node`, (contentType) => {
+  let creator = creators[contentType];
+  assert.isNotNull(creator, `I do not know how to create ${contentType} nodes yet.  Please add a definition in ${__filename}.`);
```
Comment (acrollet) [inline tests/behavioral/cypress/integration/common/i_create_a_node.js:16]: 💯

File: tests/behavioral/cypress/integration/common/i_am_logged_in_as_a_user_with_the_role.js (hunk 4 above)
Comment (ndouglas) [reply tests/behavioral/cypress/integration/common/i_am_logged_in_as_a_user_with_the_role.js:7]: Sure.  Do you think we should not worry about removing them if a test fails (to allow some freedom in debugging) but remove them after a successful test?

Comment (acrollet) [reply tests/behavioral/cypress/integration/common/i_am_logged_in_as_a_user_with_the_role.js:7]: hmm, perhaps we could detect if the tests are being run interactively and not remove in that case? The concern is that leftover test content will be confusing for end users in tugboat environments and on dev/staging. If we can move to ephemeral DBs for testing then I wouldn't be concerned at all... LMK what you think.

Comment (ndouglas) [reply tests/behavioral/cypress/integration/common/i_am_logged_in_as_a_user_with_the_role.js:7]: Or just keep 'em on Lando, zap 'em elsewhere?

Comment (acrollet) [reply tests/behavioral/cypress/integration/common/i_am_logged_in_as_a_user_with_the_role.js:7]: That works for me - this would be a good thing to revisit with everyone when we have the testing spike I think.

Comment (ndouglas) [reply tests/behavioral/cypress/integration/common/i_am_logged_in_as_a_user_with_the_role.js:7]: Added that to #4109 😃

Comment (ndouglas) [reply tests/behavioral/cypress/integration/common/i_am_logged_in_as_a_user_with_the_role.js:7]: Did some reading about this and it seems to be a nightmare.

Cypress [warns against](https://docs.cypress.io/guides/references/best-practices.html#Using-after-or-afterEach-hooks) using the after/afterEach hooks for this.  So okay, we could use before/beforeEach.  

The best use of that would be to just set the DB back to a known state, I figure.  That also removes any work performed by users or any other test processes.  Guessing that's an issue on BRD.

But ignoring that, the process would seem to be either costly in terms of time (if we just reimported the DB at the end) or simply not possible (ideally, we could just start a transaction when tests are started and then cancel it when tests end, but MySQL doesn't support nested transactions [does anything?  probably PostgreSQL... [eh, not really](https://stackoverflow.com/questions/25425944/does-postgres-support-nested-or-autonomous-transactions)].

**TL;DR**: I was hoping to throw in a few lines that'd be 99% effective pending the testing spike.  Looks like that's not going to happen.

Given that we already create junk test data, I feel fairly comfortable doing diddly about this until the spike.

Comment (acrollet) [reply tests/behavioral/cypress/integration/common/i_am_logged_in_as_a_user_with_the_role.js:7]: Ok, per earlier discussions with @swirtSJW perhaps we could prefix the auto-generated usernames/titles with `test_` or some such to help make things clear to users?

**Comment by ndouglas** [comment]
@acrollet eh?

**Comment by acrollet** [comment]
I see the JSD widget on a 403 request as expected:

Screen Shot 2021-02-08 at 1 09 59 PM
